[vllm] [2.12 regression][B200] test_batch_invariance: nondeterministic outputs 3/5 trials with FLASH_ATTN (B200 only, H100 passes)

## Summary

Under torch 2.12.0 + triton 3.7.0, vLLM's `test_v1_generation_is_deterministic_across_batch_sizes_with_needle[FLASH_ATTN]` fails on **B200** because outputs diverge across batch sizes:

> `Failed: Nondeterministic outputs detected: 3 failed out of 5 trials (max_batch_size=128).`

Job reports: `[determinism] total=5, passed=2, failed=3, max_batch_size=128`

The test generates the same prompts under different batch sizes and asserts output identity; a single test run produces identical outputs, but across batch sizes the outputs diverge. Note: **H100 passes** the same test — only **B200** fails. Passes on torch 2.11. Blocking the torch 2.12 upgrade for vLLM (vllm-project/vllm#40077).

## Environment

- `torch`: 2.12.0+cu130 (test channel)
- `triton`: 3.7.0
- CUDA: 13.0
- Python: 3.12.13
- GPU: NVIDIA B200 (fails) — NVIDIA H100 (passes the same test)
- Attention backend: `FLASH_ATTN`

## Reproduction

Failing test:
```
tests/v1/determinism/test_batch_invariance.py::test_v1_generation_is_deterministic_across_batch_sizes_with_needle[FLASH_ATTN]
```

Repro:
```
export VLLM_TEST_MODEL=Qwen/Qwen3-30B-A3B-Thinking-2507-FP8
pytest -v -s 'v1/determinism/test_batch_invariance.py::test_v1_generation_is_deterministic_across_batch_sizes_with_needle[FLASH_ATTN]'
```

The test runs 5 trials; each trial compares a single-request generation to the same prompt batched with other requests. 3 of 5 trials produced mismatching outputs on B200.

## Sample divergence

```
== Not the same as ==

100-story building. A man was standing on the 100th floor. He was holding a glass of water.
He dropped the glass, and it fell 100 floors, yet it didn't break. How is this possible?
```

(full single-vs-batched output pairs captured in job log)

## Reproducibility

- torch-2.12 branch (B200) — still reproducing on every test-PR run since 2026-04-22:
  - 2026-04-22: https://buildkite.com/vllm/ci/builds/62583#019db793-6c6b-4e37-9e80-a0663a043dfa
  - 2026-04-24: https://buildkite.com/vllm/ci/builds/62848#019dbf56-e7f6-4e63-9feb-e137a1b65df7
  - 2026-04-27: https://buildkite.com/vllm/ci/builds/63095#019dcf15-802a-456a-aae7-93c4931e983e
  - 2026-04-28: https://buildkite.com/vllm/ci/builds/63283#019dd3d1-32bb-4326-b3b4-e03529d65006
  - 2026-04-29: https://buildkite.com/vllm/ci/builds/63658#019dda93-ef28-45fe-996b-cd49d3524638
  - All five runs hit the identical signature on the same parametrization: `[determinism] total=5, passed=2, failed=3, max_batch_size=128` on `test_v1_generation_is_deterministic_across_batch_sizes_with_needle[FLASH_ATTN]`
- torch-2.12 branch (H100): ✅ passes (same commit; H100 always passes)
- main (torch 2.11) — passes on every recent B200 run:
  - 2026-04-21 daily: https://buildkite.com/vllm/ci/builds/62383
  - 2026-04-22 nightly: https://buildkite.com/vllm/ci/builds/62456
  - 2026-04-22 daily: https://buildkite.com/vllm/ci/builds/62566
  - 2026-04-23 nightly: https://buildkite.com/vllm/ci/builds/62613
  - 2026-04-24 nightly: https://buildkite.com/vllm/ci/builds/62812
  - 2026-04-29 nightly: https://buildkite.com/vllm/ci/builds/63497
  - 2026-04-29 daily: https://buildkite.com/vllm/ci/builds/63685

Prior torch-2.12 builds (62138/62232/62495) had this test blocked earlier in init by pytorch/pytorch#180905 (bmm kernel registration). Now that #180905 is fixed and closed, the test runs to completion and exposes this B200-specific non-determinism.

## Diagnosis request

The divergence is B200-only, which points to an SM-10/Blackwell-specific kernel (flash-attention variant, softmax, matmul, or rotary) producing batch-dependent numerical drift introduced by torch 2.12 or triton 3.7. Could a maintainer investigate which op path differs between B200 and H100 in this regression?

## Links

- vLLM PR: vllm-project/vllm#40077
- Umbrella: pytorch/pytorch#180899


cc @seemethere @malfet @tinglvv @nWEIdia

---

## Update 2026-04-30 (corrected): MIG hypothesis retracted; isolated to specific B200 hardware/driver

An earlier update on this issue claimed the failure was MIG-on-B200-specific. That was wrong — I conflated the MIG-sliced agent of an unrelated job (`Multi-Modal Processor` on `queue=h200_18gb`) with the agent for this `Batch Invariance (B200)` job. **Both the failing B200 agent and the passing H100 agent run on full GPUs with MIG disabled.**

### Verified agent details (from build 63658 nvidia-smi output)

**Failing CI agent** `dgxB200-09-3` (queue=B200):
```
NVIDIA-SMI 580.65.06   Driver 580.65.06   CUDA 13.0
NVIDIA B200, 183359 MiB, MIG M.: Disabled    (full GPU, both GPU 0 and GPU 1)
```

**Passing CI agent** `mithril-h100-pool`:
```
NVIDIA-SMI 580.95.05   Driver 580.95.05   CUDA 13.0
NVIDIA H100 80GB HBM3, 81559 MiB, MIG M.: Disabled    (full GPU)
```

### Reproduction on a full B200 dev box (also MIG disabled)

```
NVIDIA-SMI 595.58.03   Driver 595.58.03   CUDA 13.2
NVIDIA B200, 183359 MiB, MIG M.: Disabled    (full GPU)
```

Same vLLM commit (`1e8d2aca9` from build 63658), same torch wheel (`torch==2.12.0+cu130`, test channel 2026-04-28 23:18 RC), same `triton==3.7.0`, same `[FLASH_ATTN]` parametrization.

| Field | CI `dgxB200-09-3` (❌ fails) | Dev box mine (✅ passes) | CI `mithril-h100-pool` (✅ passes) |
|---|---|---|---|
| System | **NVIDIA DGX B200** | **AWS EC2 `p6-b200.48xlarge`** | k8s pool with H100 |
| GPU model | NVIDIA B200 | NVIDIA B200 | NVIDIA H100 80GB HBM3 |
| GPU memory | 183 GiB | 183 GiB | 80 GiB |
| MIG | Disabled (full GPU) | Disabled (full GPU) | Disabled (full GPU) |
| Driver | **580.65.06** | **595.58.03** | 580.95.05 |
| CUDA (host) | 13.0 | 13.2 | 13.0 |
| GPU UUID | (not captured in default `nvidia-smi` log) | `GPU-eff083f7-0e58-04bf-0f2d-a26dcef6669b` | (not captured) |
| GPU VBIOS | (not captured) | `97.00.C5.00.2D` | (not captured) |
| Result | `passed=2, failed=3` deterministically (5 consecutive runs Apr 22–29) | `passed=5, failed=0` (6 consecutive runs Apr 30 — 3× at `mem_util=0.5`, 3× at `mem_util=0.9`) | passes every run |

So the failure is **not** caused by MIG, **not** by SM count, **not** by memory pressure (`gpu_memory_utilization=0.5` and `0.9` both clean on dev box), and **not** by torch/triton/vLLM versions (identical across all three environments).

### What's actually different

1. **Platform**: CI runs on an **NVIDIA DGX B200** reference system (`dgxB200-09-3`). My passing dev box is an **AWS EC2 `p6-b200.48xlarge`**. These are different OEM platforms with different VBIOS, NVSwitch/NVLink fabric topology, NUMA layout, and PCIe lane allocation — even though the Blackwell die itself is the same.
2. **Host driver version**: CI is on **580.65.06 / CUDA 13.0**, dev box is on **595.58.03 / CUDA 13.2**. The 595 series is roughly 6 months newer; it may include kernel-launch-ordering, UVM, or stream-synchronization fixes that affect the floating-point reduction order in FlashAttention v2.
3. **Specific hardware unit**: even within DGX B200, `dgxB200-09-3` is a single physical node. Possible per-node variation: VBIOS revision (CI's VBIOS isn't captured by the default `nvidia-smi` block in the log), ECC retired pages, persistent SM clock policy, etc.

### Diagnosis request (corrected)

This now looks like an **environment-specific regression** tied to the DGX B200 platform and/or driver 580.65.06 — the same vLLM/torch wheel produces deterministic outputs on AWS p6-b200 with driver 595.58.03. Could a maintainer:

1. Confirm whether driver `580.65.06` has any known FlashAttention v2 / Blackwell numerical issues that were fixed in the `595.x` series.
2. Reproduce on **another DGX B200** (different physical unit) — if every dgxB200 reproduces, it's a driver/firmware issue. If only `dgxB200-09-3` reproduces, it's per-unit hardware state.
3. **Capture VBIOS/UUID in CI** by adding `nvidia-smi --query-gpu=vbios_version,uuid,serial --format=csv` to the test setup script, so we can compare exactly which DGX firmware vs which AWS firmware behaves how.

A targeted repro inside `dgxB200-09-3` with driver upgraded to 595.x (matching the AWS dev box) would tell us whether the driver alone is the culprit.

Here is the repro:

```
====================================================================================== test session starts =======================================================================================
platform linux -- Python 3.12.13, pytest-8.3.5, pluggy-1.5.0 -- /usr/bin/python3
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase(PosixPath('/vllm-workspace/tests/.hypothesis/examples'))
rootdir: /vllm-workspace
configfile: pyproject.toml
plugins: asyncio-0.24.0, timeout-2.3.1, hypothesis-6.131.0, anyio-4.6.2.post1, mock-3.14.0, cov-6.3.0, rerunfailures-14.0, forked-1.6.0, schemathesis-3.39.15, hydra-core-1.3.2, subtests-0.14.1, shard-0.1.2, buildkite-test-collector-0.1.9
asyncio: mode=Mode.STRICT, default_loop_scope=None
collected 1 item                                                                                                                                                                                 
Running 1 items in this shard: tests/v1/determinism/test_batch_invariance.py::test_v1_generation_is_deterministic_across_batch_sizes_with_needle[FLASH_ATTN]

v1/determinism/test_batch_invariance.py::test_v1_generation_is_deterministic_across_batch_sizes_with_needle[FLASH_ATTN] INFO 04-30 22:49:20 [utils.py:233] non-default args: {'max_model_len': 5120, 'enable_prefix_caching': False, 'gpu_memory_utilization': 0.5, 'max_num_seqs': 128, 'disable_log_stats': True, 'attention_config': AttentionConfig(backend=<AttentionBackendEnum.FLASH_ATTN: 'vllm.v1.attention.backends.flash_attn.FlashAttentionBackend'>, flash_attn_version=None, use_prefill_decode_attention=False, flash_attn_max_num_splits_for_cuda_graph=32, tq_max_kv_splits_for_cuda_graph=32, use_cudnn_prefill=False, use_trtllm_ragged_deepseek_prefill=False, use_trtllm_attention=None, disable_flashinfer_prefill=True, disable_flashinfer_q_quantization=False, use_prefill_query_quantization=False, use_fp4_indexer_cache=False, use_non_causal=False), 'model': 'Qwen/Qwen3-30B-A3B-Thinking-2507-FP8'}
WARNING 04-30 22:49:20 [envs.py:1820] Unknown vLLM environment variable detected: VLLM_TEST_MODEL
INFO 04-30 22:49:21 [model.py:563] Resolved architecture: Qwen3MoeForCausalLM
INFO 04-30 22:49:21 [model.py:1692] Using max model len 5120
INFO 04-30 22:49:21 [nixl_utils.py:20] Setting UCX_RCACHE_MAX_UNRELEASED to '1024' to avoid a rare memory leak in UCX when using NIXL.
WARNING 04-30 22:49:21 [nixl_utils.py:34] NIXL is not available
WARNING 04-30 22:49:21 [nixl_utils.py:44] NIXL agent config is not available
INFO 04-30 22:49:22 [scheduler.py:239] Chunked prefill is enabled with max_num_batched_tokens=16384.
INFO 04-30 22:49:22 [vllm.py:840] Asynchronous scheduling is enabled.
INFO 04-30 22:49:22 [kernel.py:205] Final IR op priority after setting platform defaults: IrOpPriorityConfig(rms_norm=['native'])
INFO 04-30 22:49:22 [compilation.py:303] Enabled custom fusions: norm_quant, act_quant
(EngineCore pid=9216) INFO 04-30 22:49:23 [core.py:109] Initializing a V1 LLM engine (v0.20.1rc1.dev77+g1e8d2aca9) with config: model='Qwen/Qwen3-30B-A3B-Thinking-2507-FP8', speculative_config=None, tokenizer='Qwen/Qwen3-30B-A3B-Thinking-2507-FP8', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.bfloat16, max_seq_len=5120, download_dir=None, load_format=auto, tensor_parallel_size=1, pipeline_parallel_size=1, data_parallel_size=1, decode_context_parallel_size=1, dcp_comm_backend=ag_rs, disable_custom_all_reduce=True, quantization=fp8, quantization_config=None, enforce_eager=False, enable_return_routed_experts=False, kv_cache_dtype=auto, device_config=cuda, structured_outputs_config=StructuredOutputsConfig(backend='auto', disable_any_whitespace=False, disable_additional_properties=False, reasoning_parser='', reasoning_parser_plugin='', enable_in_reasoning=False), observability_config=ObservabilityConfig(show_hidden_metrics_for_version=None, otlp_traces_endpoint=None, collect_detailed_traces=None, kv_cache_metrics=False, kv_cache_metrics_sample=0.01, cudagraph_metrics=False, enable_layerwise_nvtx_tracing=False, enable_mfu_metrics=False, enable_mm_processor_stats=False, enable_logging_iteration_details=False), seed=0, served_model_name=Qwen/Qwen3-30B-A3B-Thinking-2507-FP8, enable_prefix_caching=False, enable_chunked_prefill=True, pooler_config=None, compilation_config={'mode': <CompilationMode.VLLM_COMPILE: 3>, 'debug_dump_path': None, 'cache_dir': '', 'compile_cache_save_format': 'binary', 'backend': 'inductor', 'custom_ops': ['+quant_fp8', 'none', '+quant_fp8'], 'ir_enable_torch_wrap': True, 'splitting_ops': ['vllm::unified_attention_with_output', 'vllm::unified_mla_attention_with_output', 'vllm::mamba_mixer2', 'vllm::mamba_mixer', 'vllm::short_conv', 'vllm::linear_attention', 'vllm::plamo2_mamba_mixer', 'vllm::gdn_attention_core', 'vllm::gdn_attention_core_xpu', 'vllm::olmo_hybrid_gdn_full_forward', 'vllm::kda_attention', 'vllm::sparse_attn_indexer', 'vllm::rocm_aiter_sparse_attn_indexer', 'vllm::deepseek_v4_attention', 'vllm::unified_kv_cache_update', 'vllm::unified_mla_kv_cache_update'], 'compile_mm_encoder': False, 'cudagraph_mm_encoder': False, 'encoder_cudagraph_token_budgets': [], 'encoder_cudagraph_max_vision_items_per_batch': 0, 'encoder_cudagraph_max_frames_per_batch': None, 'compile_sizes': [], 'compile_ranges_endpoints': [16384], 'inductor_compile_config': {'enable_auto_functionalized_v2': False, 'combo_kernels': True, 'benchmark_combo_kernel': True}, 'inductor_passes': {}, 'cudagraph_mode': <CUDAGraphMode.FULL_AND_PIECEWISE: (2, 1)>, 'cudagraph_num_of_warmups': 1, 'cudagraph_capture_sizes': [1, 2, 4, 8, 16, 24, 32, 40, 48, 56, 64, 72, 80, 88, 96, 104, 112, 120, 128, 136, 144, 152, 160, 168, 176, 184, 192, 200, 208, 216, 224, 232, 240, 248, 256], 'cudagraph_copy_inputs': False, 'cudagraph_specialize_lora': True, 'use_inductor_graph_partition': False, 'pass_config': {'fuse_norm_quant': True, 'fuse_act_quant': True, 'fuse_attn_quant': False, 'enable_sp': False, 'fuse_gemm_comms': False, 'fuse_allreduce_rms': False}, 'max_cudagraph_capture_size': 256, 'dynamic_shapes_config': {'type': <DynamicShapesType.BACKED: 'backed'>, 'evaluate_guards': False, 'assume_32_bit_indexing': False}, 'local_cache_dir': None, 'fast_moe_cold_start': False, 'static_all_moe_layers': []}, kernel_config=KernelConfig(ir_op_priority=IrOpPriorityConfig(rms_norm=['native']), enable_flashinfer_autotune=True, moe_backend='auto')
[W430 22:49:24.498390193 Context.cpp:541] Warning: torch.backends.cuda.preferred_blas_library is an experimental feature. If you see any error or unexpected behavior when this flag is set please file an issue on GitHub. (function operator())
(EngineCore pid=9216) INFO 04-30 22:49:24 [parallel_state.py:1402] world_size=1 rank=0 local_rank=0 distributed_init_method=tcp://172.17.0.2:40355 backend=nccl
(EngineCore pid=9216) INFO 04-30 22:49:24 [parallel_state.py:1715] rank 0 in world size 1 is assigned as DP rank 0, PP rank 0, PCP rank 0, TP rank 0, EP rank 0, EPLB rank N/A
(EngineCore pid=9216) INFO 04-30 22:49:25 [topk_topp_sampler.py:45] Using FlashInfer for top-p & top-k sampling.
(EngineCore pid=9216) INFO 04-30 22:49:25 [gpu_model_runner.py:4781] Starting to load model Qwen/Qwen3-30B-A3B-Thinking-2507-FP8...
(EngineCore pid=9216) INFO 04-30 22:49:25 [__init__.py:389] Selected DeepGemmFp8BlockScaledMMKernel for Fp8LinearMethod
(EngineCore pid=9216) INFO 04-30 22:49:25 [deep_gemm.py:113] DeepGEMM E8M0 enabled on current platform.
(EngineCore pid=9216) WARNING 04-30 22:49:25 [fa_utils.py:138] Cannot use FA version 4 with batch invariance, defaulting to FA version 2.
(EngineCore pid=9216) INFO 04-30 22:49:25 [cuda.py:308] Using AttentionBackendEnum.FLASH_ATTN backend.
(EngineCore pid=9216) INFO 04-30 22:49:25 [flash_attn.py:646] Using FlashAttention version 2
(EngineCore pid=9216) INFO 04-30 22:49:25 [fp8.py:394] Using TRITON Fp8 MoE backend out of potential backends: ['AITER', 'FLASHINFER_TRTLLM', 'FLASHINFER_CUTLASS', 'DEEPGEMM', 'TRITON', 'MARLIN', 'BATCHED_DEEPGEMM', 'BATCHED_TRITON', 'XPU'].
(EngineCore pid=9216) INFO 04-30 22:49:26 [weight_utils.py:904] Filesystem type for checkpoints: OVERLAY. Checkpoint size: 29.03 GiB. Available RAM: 1917.16 GiB.
(EngineCore pid=9216) INFO 04-30 22:49:26 [weight_utils.py:927] Auto-prefetch is disabled because the filesystem (OVERLAY) is not a recognized network FS (NFS/Lustre). If you want to force prefetching, start vLLM with --safetensors-load-strategy=prefetch.
Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]
Loading safetensors checkpoint shards:  25% Completed | 1/4 [00:02<00:06,  2.19s/it]
Loading safetensors checkpoint shards:  50% Completed | 2/4 [00:05<00:05,  2.65s/it]
Loading safetensors checkpoint shards:  75% Completed | 3/4 [00:07<00:02,  2.53s/it]
Loading safetensors checkpoint shards: 100% Completed | 4/4 [00:07<00:00,  1.66s/it]
Loading safetensors checkpoint shards: 100% Completed | 4/4 [00:07<00:00,  1.97s/it]
(EngineCore pid=9216) 
(EngineCore pid=9216) INFO 04-30 22:49:34 [default_loader.py:391] Loading weights took 7.90 seconds
(EngineCore pid=9216) INFO 04-30 22:49:35 [fp8.py:578] Using MoEPrepareAndFinalizeNoDPEPModular
(EngineCore pid=9216) INFO 04-30 22:49:35 [gpu_model_runner.py:4883] Model loading took 29.13 GiB memory and 10.058557 seconds
(EngineCore pid=9216) INFO 04-30 22:49:44 [backends.py:1070] Using cache directory: /root/.cache/vllm/torch_compile_cache/372c124628/rank_0_0/backbone for vLLM's torch.compile
(EngineCore pid=9216) INFO 04-30 22:49:44 [backends.py:1129] Dynamo bytecode transform time: 8.63 s
(EngineCore pid=9216) INFO 04-30 22:49:46 [backends.py:291] Directly load the compiled graph(s) for compile range (1, 16384) from the cache, took 0.431 s
(EngineCore pid=9216) INFO 04-30 22:49:47 [monitor.py:53] torch.compile and initial profiling/warmup run together took 11.26 s in total
(EngineCore pid=9216) INFO 04-30 22:49:47 [gpu_model_runner.py:5967] Profiling CUDA graph memory: PIECEWISE=35 (largest=256), FULL=19 (largest=128)
(EngineCore pid=9216) INFO 04-30 22:49:48 [gpu_model_runner.py:6046] Estimated CUDA graph memory: 0.34 GiB total
(EngineCore pid=9216) INFO 04-30 22:49:49 [gpu_worker.py:433] Available KV cache memory: 54.86 GiB
(EngineCore pid=9216) INFO 04-30 22:49:49 [gpu_worker.py:448] CUDA graph memory profiling is enabled (default since v0.21.0). The current --gpu-memory-utilization=0.5000 is equivalent to --gpu-memory-utilization=0.4981 without CUDA graph memory profiling. To maintain the same effective KV cache size as before, increase --gpu-memory-utilization to 0.5019. To disable, set VLLM_MEMORY_PROFILER_ESTIMATE_CUDAGRAPHS=0.
(EngineCore pid=9216) INFO 04-30 22:49:49 [kv_cache_utils.py:1710] GPU KV cache size: 599,152 tokens
(EngineCore pid=9216) INFO 04-30 22:49:49 [kv_cache_utils.py:1711] Maximum concurrency for 5,120 tokens per request: 117.02x
DeepGEMM warmup: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 841/841 [00:00<00:00, 22083.30it/s]
(EngineCore pid=9216) 2026-04-30 22:49:49,162 - INFO - autotuner.py:457 - flashinfer.jit: [Autotuner]: Autotuning process starts ...
(EngineCore pid=9216) 2026-04-30 22:49:49,205 - INFO - autotuner.py:466 - flashinfer.jit: [Autotuner]: Autotuning process ends
Capturing CUDA graphs (mixed prefill-decode, PIECEWISE): 100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 35/35 [00:03<00:00, 11.63it/s]
Capturing CUDA graphs (decode, FULL): 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 19/19 [00:01<00:00, 13.39it/s]
(EngineCore pid=9216) INFO 04-30 22:49:54 [gpu_model_runner.py:6137] Graph capturing finished in 5 secs, took -5.45 GiB
(EngineCore pid=9216) INFO 04-30 22:49:54 [gpu_worker.py:592] CUDA graph pool memory: -5.45 GiB (actual), 0.34 GiB (estimated), difference: 5.79 GiB (621176422400.0%).
(EngineCore pid=9216) INFO 04-30 22:49:54 [core.py:299] init engine (profile, create kv cache, warmup model) took 18.70 s (compilation: 11.26 s)
(EngineCore pid=9216) INFO 04-30 22:49:54 [kernel.py:205] Final IR op priority after setting platform defaults: IrOpPriorityConfig(rms_norm=['native'])
Rendering prompts: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 40.47it/s]
Processed prompts: 100%|██████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.20it/s, est. speed input: 6.02 toks/s, output: 154.17 toks/s]
Rendering prompts: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 117/117 [00:00<00:00, 1800.76it/s]
Processed prompts: 100%|██████████████████████████████████████████████████████████████████████████████| 117/117 [00:04<00:00, 26.86it/s, est. speed input: 8946.78 toks/s, output: 3438.29 toks/s]
100-story building. A 100-story building is a building with 100 floors. A 100-story building is a building with 100 floors. A 100-story building is a building with 100 floors. A 100-story building is a building with 100 floors. A 100-story building is a building with 100 floors. A 100-story building is a building with 100 floors. A 100-story building is a building with 100 floors. A 1

== Not the same as ==

100-story building. A man was standing on the 100th floor. He was holding a glass of water. He dropped the glass, and it fell 100 floors, yet it didn't break. How is this possible?

 (A) The glass was empty.
 (B) The building was on the ground floor.
 (C) The glass was made of rubber.
 (D) The man was on the first floor.

I need to figure out this riddle. There's a 100-story building, and a man is on the 100th floor holding a glass of water


Rendering prompts: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 126/126 [00:00<00:00, 1826.11it/s]
Processed prompts: 100%|█████████████████████████████████████████████████████████████████████████████| 126/126 [00:03<00:00, 36.25it/s, est. speed input: 12289.13 toks/s, output: 4639.92 toks/s]
100-story building. A 100-story building is a building with 100 floors. A 100-story building is a building with 100 floors. A 100-story building is a building with 100 floors. A 100-story building is a building with 100 floors. A 100-story building is a building with 100 floors. A 100-story building is a building with 100 floors. A 100-story building is a building with 100 floors. A 1

== Not the same as ==

100-story building. A man was standing on the 100th floor. He was holding a glass of water. He dropped the glass, and it fell 100 floors, yet it didn't break. How is this possible?

 (A) The glass was empty.
 (B) The building was on the ground floor.
 (C) The glass was made of rubber.
 (D) The man was on the first floor.

I need to figure out this riddle. There's a 100-story building, and a man is on the 100th floor holding a glass of water


Rendering prompts: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 117/117 [00:00<00:00, 1926.66it/s]
Processed prompts: 100%|█████████████████████████████████████████████████████████████████████████████| 117/117 [00:03<00:00, 34.98it/s, est. speed input: 11186.93 toks/s, output: 4477.88 toks/s]
Rendering prompts: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 113/113 [00:00<00:00, 1828.75it/s]
Processed prompts: 100%|█████████████████████████████████████████████████████████████████████████████| 113/113 [00:03<00:00, 32.99it/s, est. speed input: 11325.68 toks/s, output: 4222.79 toks/s]
Rendering prompts: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 86/86 [00:00<00:00, 1886.48it/s]
Processed prompts: 100%|████████████████████████████████████████████████████████████████████████████████| 86/86 [00:03<00:00, 28.04it/s, est. speed input: 9238.71 toks/s, output: 3588.93 toks/s]
100-story building. A man was standing on the 50th floor. He was holding a 100-pound weight. He dropped the weight from the window. The weight hit the ground and broke into 100 pieces. How did he get down?

I need to figure out this puzzle. It says there's a 100-story building, and a man is on the 50th floor holding a 100-pound weight. He drops the weight from the window, and it hits the ground and breaks into 100 pieces. Then it asks how he got down.

First

== Not the same as ==

100-story building. A man was standing on the 100th floor. He was holding a glass of water. He dropped the glass, and it fell 100 floors, yet it didn't break. How is this possible?

 (A) The glass was empty.
 (B) The building was on the ground floor.
 (C) The glass was made of rubber.
 (D) The man was on the first floor.

I need to figure out this riddle. There's a 100-story building, and a man is on the 100th floor holding a glass of water


[determinism] total=5, passed=2, failed=3, max_batch_size=128
FAILED(EngineCore pid=9216) INFO 04-30 22:50:14 [core.py:1238] Shutdown initiated (timeout=0)
(EngineCore pid=9216) INFO 04-30 22:50:14 [core.py:1261] Shutdown complete


============================================================================================ FAILURES ============================================================================================
_________________________________________________________ test_v1_generation_is_deterministic_across_batch_sizes_with_needle[FLASH_ATTN] _________________________________________________________

backend = 'FLASH_ATTN'

    @skip_unsupported
    @pytest.mark.timeout(1000)
    @pytest.mark.parametrize(
        "backend",
        BACKENDS,
    )
    def test_v1_generation_is_deterministic_across_batch_sizes_with_needle(
        backend,
    ):
        """
        Ensures that the same request (the 'needle' prompt) yields identical output
        whether run alone (bs=1) or mixed into a larger batch (e.g., bs=64),
        using the high-level v1 LLM() API only (no manual batching).
    
        Strategy:
        - Create a single LLM engine configured for the larger batch limit (N).
        - Compute a baseline output for the needle prompt when it is run alone.
        - For many trials, generate a mixed batch (size N) where the needle appears
          at a random position among random filler prompts using the same engine.
        - Track how many trials match vs mismatch, and report totals at the end.
          The test fails if any mismatches occur, but we still dump pass/fail
          counts.
    
        Notes:
        - Use seeded stochastic sampling with a fixed seed to test determinism.
        - Outputs are intentionally longer and sampled at higher temperature/top_p
          to produce a more random-sounding phrase, yet remain deterministic by
          seed.
        - Keep max_tokens and max_model_len bounded for speed and memory use.
        """
        seed = int(os.getenv("VLLM_TEST_SEED", "12345"))
        random.seed(seed)
    
        attention_config = {"backend": backend}
        # Allow overrides from environment (useful for CI tuning)
        # "facebook/opt-125m" is too small, doesn't reliably test determinism
        model = TEST_MODEL
        num_trials = int(os.getenv("VLLM_NEEDLE_TRIALS", "5"))
        max_batch_size = int(os.getenv("VLLM_NEEDLE_BATCH_SIZE", "128"))
        min_random_prompt = int(os.getenv("VLLM_MIN_PROMPT", "1024"))
        max_random_prompt = int(os.getenv("VLLM_MAX_PROMPT", "2048"))
        assert max_batch_size >= 2, "Batch size should be >= 2 to mix needle."
    
        # Keep GPU memory usage low to avoid startup allocation failures.
        gpu_mem_util = float(os.getenv("VLLM_GPU_MEMORY_UTILIZATION", "0.5"))
        max_model_len = int(os.getenv("VLLM_MAX_MODEL_LEN", "5120"))
    
        # Sampling parameters: longer outputs with a more random-sounding
        # continuation,but still deterministic due to fixed seed.
        temperature = float(os.getenv("VLLM_NEEDLE_TEMPERATURE", "0.0"))
        top_p = float(os.getenv("VLLM_NEEDLE_TOP_P", "0.95"))
        max_tokens = int(os.getenv("VLLM_NEEDLE_MAX_TOKENS", "128"))
    
        sampling = SamplingParams(
            temperature=temperature,
            top_p=top_p,
            max_tokens=max_tokens,
            seed=20240919,
        )
    
        needle_prompt = "There once was a "
    
        llm = None
        try:
            llm = LLM_with_max_seqs(
                model=model,
                max_num_seqs=max_batch_size,
                gpu_memory_utilization=gpu_mem_util,
                max_model_len=max_model_len,
                attention_config=attention_config,
            )
    
            # Baseline generation for the needle prompt alone.
            baseline_out = llm.generate([needle_prompt], sampling)
            assert len(baseline_out) == 1
            assert len(baseline_out[0].outputs) >= 1
            baseline_text = baseline_out[0].outputs[0].text
    
            mismatches = 0
    
            for trial in range(num_trials):
                # Create a batch of size `max_batch_size` and insert the needle at
                # a random index
                prompts: list[str] = []
                batch_size = random.randint(max_batch_size // 2, max_batch_size)
                needle_pos = random.randint(0, batch_size - 1)
                for i in range(batch_size):
                    if i == needle_pos:
                        prompts.append(needle_prompt)
                    else:
                        prompts.append(_random_prompt(min_random_prompt, max_random_prompt))
    
                # Generate with the same engine but in a larger batch.
                outputs = llm.generate(prompts, sampling)
                # Find the needle output by position
                needle_output = outputs[needle_pos]
                assert needle_output.prompt == needle_prompt
                assert len(needle_output.outputs) >= 1
                text = needle_output.outputs[0].text
    
                if text != baseline_text:
                    print(f"{text}\n\n== Not the same as ==\n\n{baseline_text}\n\n")
                    mismatches += 1
    
            passes = num_trials - mismatches
            # Dump how many passed vs failed
            print(
                f"[determinism] total={num_trials}, passed={passes}, "
                f"failed={mismatches}, max_batch_size={max_batch_size}"
            )
    
            if mismatches > 0:
>               pytest.fail(
                    f"Nondeterministic outputs detected: {mismatches} failed out "
                    f"of {num_trials} trials (max_batch_size={max_batch_size})."
                )
E               Failed: Nondeterministic outputs detected: 3 failed out of 5 trials (max_batch_size=128).

v1/determinism/test_batch_invariance.py:136: Failed
======================================================================================== warnings summary ========================================================================================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

../../usr/local/lib/python3.12/dist-packages/schemathesis/generation/coverage.py:305
  /usr/local/lib/python3.12/dist-packages/schemathesis/generation/coverage.py:305: DeprecationWarning: jsonschema.exceptions.RefResolutionError is deprecated as of version 4.18.0. If you wish to catch potential reference resolution errors, directly catch referencing.exceptions.Unresolvable.
    ref_error: type[Exception] = jsonschema.RefResolutionError,

tests/v1/determinism/test_batch_invariance.py: 14 warnings
  /usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:365: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

tests/v1/determinism/test_batch_invariance.py::test_v1_generation_is_deterministic_across_batch_sizes_with_needle[FLASH_ATTN]
  /usr/lib/python3.12/multiprocessing/popen_fork.py:66: DeprecationWarning: This process (pid=8948) is multi-threaded, use of fork() may lead to deadlocks in the child.
    self.pid = os.fork()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
==================================================================================== short test summary info =====================================================================================
FAILED v1/determinism/test_batch_invariance.py::test_v1_generation_is_deterministic_across_batch_sizes_with_needle[FLASH_ATTN] - Failed: Nondeterministic outputs detected: 3 failed out of 5 trials (max_batch_size=128).
================================================================================ 1 failed, 18 warnings in 54.45s =================================================================================
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute
```

Reopening — the same signature reproduces on a fresh test-PR build with current torch nightly.

## Reproduction on torch 2.12 / vLLM PR #42848

- **2026-05-20** — Buildkite build [#66993](https://buildkite.com/vllm/ci/builds/66993#019e4286-21ff-4236-8492-7fd3543df34b) on rebased commit `d2792bf20`:
  ```
  FAILED v1/determinism/test_batch_invariance.py::test_v1_generation_is_deterministic_across_batch_sizes_with_needle[FLASH_ATTN]
    Failed: Nondeterministic outputs detected: 4 failed out of 5 trials (max_batch_size=128).
  ```
- Same `[FLASH_ATTN]` parameterization, same B200 hardware, same failure mode as the original report.
- Not present on main #66975, #66835, #66759 (torch 2.11). Main #67062 fails the *same* job but on a different test (`test_logprobs_without_batch_invariance_should_fail` with `TypeError: fmax() missing 1 required positional argument: 'b'`), which is unrelated.

The fix that closed this issue either didn't make it into the test-channel torch wheel that vLLM CI is pulling, or has regressed. CC the umbrella: pytorch/pytorch#180899.

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -1,24 +1,23 @@
-{% set version = "3.7.0" %}
+{% set version = "3.7.1" %}
 
 package:
   name: triton
   version: {{ version }}
 
 source:
-  # Use the GitHub auto-archive instead of the official release tarball:
-  # triton-3.7.0.tar.gz (releases/download/...) is missing examples/plugins/,
-  # which CMakeLists.txt:342 references (add_subdirectory). The auto-archive
-  # contains the full repo tree. Triton has no real git submodules (backends
-  # nvidia/amd/proton are in-tree), so auto-archive works as the source.
+  # 3.7.1 added MANIFEST.in coverage for examples/ in the official release
+  # tarball (#10350), but stay on the GitHub auto-archive for consistency with
+  # the 3.7.0 recipe — Triton has no real git submodules (backends nvidia/amd/
+  # proton are in-tree), so auto-archive works as the source.
   url: https://github.com/triton-lang/triton/archive/refs/tags/v{{ version }}.tar.gz
-  sha256: ff5cdcdd1b3f30a5db06c4e690ac73b16342c8181820e2cdf74e8514065ea7b8
+  sha256: 7d998625d1035ac496d06a81117727647169422ee67b8889609f06fd5367c491
   patches:
     - patches/0001-Remove-Werror-that-cause-false-positive-build-failur.patch
     - patches/0003-Use-system-PATH-to-find-tools-in-CONDA_PREFIX.patch
     - patches/0004-Add-conda-forge-include-dirs-to-list-of-include-dirs.patch
 
 build:
-  number: 1
+  number: 0
   # TODO: windows support should be available from next version;
   #       CPU-only support still under development
   skip: true  # [win or cuda_compiler_version == "None"]

